tapdb/universe_federation_test.go

```diff
@@ -112,6 +112,192 @@ func TestUniverseFederationCRUD(t *testing.T) {
+
+	// Check that the complete log entry is as expected.
+	completeEntry := completeLogEntries[0]
+
+	require.Equal(t, servers[0], completeEntry.ServerAddr)
+	require.Equal(
+		t, universe.ProofSyncStatusComplete, completeEntry.SyncStatus,
+	)
+	require.Equal(
+		t, universe.SyncDirectionPush, completeEntry.SyncDirection,
+	)
+	require.Equal(t, uniId.String(), completeEntry.UniID.String())
+	require.Equal(t, int64(0), completeEntry.AttemptCounter)
+
+	assertProofSyncLogLeafKey(t, uniProof.LeafKey, completeEntry.LeafKey)
+	assertProofSyncLogLeaf(t, *uniProof.Leaf, completeEntry.Leaf)
+
+	// Increment the attempt counter for one of the pending log entries.
+	_, err = fedStore.UpsertFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, servers[1],
+		universe.SyncDirectionPush, universe.ProofSyncStatusPending,
+		true,
+	)
+	require.NoError(t, err)
+
+	// Check that the attempt counter was incremented as expected.
+	pendingLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusPending,
+	)
+	require.NoError(t, err)
+	require.Len(t, pendingLogEntries, 2)
+
+	for i := range pendingLogEntries {
+		entry := pendingLogEntries[i]
+		if entry.ServerAddr == servers[1] {
+			require.Equal(t, int64(1), entry.AttemptCounter)
+		} else {
+			require.Equal(t, int64(0), entry.AttemptCounter)
+		}
+	}
+
+	// Set the sync status to complete for one of the pending log entries.
+	_, err = fedStore.UpsertFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, servers[1],
+		universe.SyncDirectionPush, universe.ProofSyncStatusComplete,
+		false,
+	)
+	require.NoError(t, err)
+
+	// Check that the sync status was updated as expected.
+	pendingLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusPending,
+	)
+	require.NoError(t, err)
+	require.Len(t, pendingLogEntries, 1)
+
+	completeLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusComplete,
+	)
+	require.NoError(t, err)
+	require.Len(t, completeLogEntries, 2)
+
+	// Delete log entries for one of the servers.
+	err = fedStore.DeleteProofsSyncLogEntries(ctx, servers[0], servers[1])
+	require.NoError(t, err)
+
+	// Only one log entry should remain and it should have sync status
+	// pending.
+	pendingLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusPending,
+	)
+	require.NoError(t, err)
+	require.Len(t, pendingLogEntries, 1)
+
+	// Check that the remaining log entry is as expected.
+	pendingEntry := pendingLogEntries[0]
+	require.Equal(t, servers[2], pendingEntry.ServerAddr)
+}
+
+func assertProofSyncLogLeafKey(t *testing.T, actualLeafKey universe.LeafKey,
+	logLeafKey universe.LeafKey) {
+
+	// TODO(ffranr): Can we really ignore the tweaked script key?
```

Comment:
Ah, I wasn't sure if those were questions to the reviewer or just things you wanted to figure out before finalizing the PR.

> // TODO(ffranr): Can we really ignore the tweaked script key?

Yes, since the tweaked key is the derivation information of the script key, that is only ever known to the "owner" of an asset and is never serialized in a proof. So it is correct to blank it out before comparing it here.

> // TODO(ffranr): Do we need to check all the proof leaf fields?

Yes, I think it makes sense to also compare the other fields. Does something fail if we just do `require.Equal(t, actualLeaf, logLeaf)`?

Reply:
When it comes to comparing proof leaf fields: the asset (decoded from the leaf genesis proof) does not contain the correct `GroupKey.Witness`.

Reply:
Ah, that can also be set to `nil`, as it's basically just extracted from the asset witness (which probably only happens at DB level and not when parsing a proof). So that's okay to blank out before comparison.